include/eve/module/core/regular/impl/simd/arm/sve/first_true.hpp

```diff
@@ -22,35 +22,44 @@ namespace eve::detail
     if constexpr (C::is_complete && !C::is_inverted) return std::nullopt;
     else
     {
-      L c_m = m;
-
-      // Compute the condition mask only if necessary, this gives slightly better codegen.
-      // This also masks the inactive lanes of the input.
-      if constexpr (!C::is_complete || !std::same_as<N, expected_cardinal_t<T>>)
-      {
-        if constexpr (relative_conditional_expr<C>) c_m = sve_true(cx, as(m));
-        else                                        c_m = expand_mask(cx, as<L>{});
-      }
-
       if constexpr (has_aggregated_abi_v<L>)
       {
-        if constexpr (!C::is_complete) m = m && c_m;
+        if constexpr (!C::is_complete)
+        {
+          if constexpr (relative_conditional_expr<C>) m = m && sve_true(cx, as(m));
+          else                                        m = m && expand_mask(cx, as<L>{});
+        }
 
         auto [lo, hi] = m.slice();
         auto lo_res   = first_true(lo);
         auto hi_res   = first_true(hi);
-        
+
         if (lo_res) return lo_res;
         if (hi_res) *hi_res += lo.size();
 
         return hi_res;
       }
       else
       {
+        L c_m = m;
+
+        const auto is_ec = std::same_as<N, expected_cardinal_t<T>>;
+
+        // Compute the condition mask only if necessary, this gives slightly better codegen.
+        // This also masks the inactive lanes of the input.
+        if constexpr (!C::is_complete || !is_ec)
+        {
+          if      constexpr (relative_conditional_expr<C>) c_m = sve_true(cx, as(m));
+          else if constexpr (is_ec)                        c_m = expand_mask(cx, as<L>{});
+          // merge everything right away in this case to eliminate one and instruction
+          else                                             c_m = m && remove_garbage(expand_mask(cx, as<L>{}));
+        }
```

Comment:
I appreciate that you written this in an older pr, but this took me a while to navigate, so let's fix it.

>Compute the condition mask only if necessary,

Not helpful comment. I guess you mean don't call `expand_mask`.

> merge everything right away in this case to eliminate one and instruction

I don't get it. m && here definetly seems wrong - you want to and them after svptest_any§

> remove_garbage

remove garbage here is clearly wrong. Because if you have no conditional, you forgot to remove garbage from the `m` and garbage just slips through.